### ISSUE TITLE
LSP: Improve the signature help by highlighting the current active parameter

### ISSUE THREAD

**Issue by urandom** [neovim/neovim#14444]
This might be a bit tricky, but it would be nice if the signature help is improved by somehow highlighting the currently active parameter (with bold/different color/etc). Perhaps it would be possible to extend  `util.convert_signature_help_to_markdown_lines` so that it changes the signatures[activeSignature].label to wrap the activeParameter with bold `**` markdown symbols, if those are then handled by the popup.


The following excerpt is from gopls, go's language server. It might be possible to use `parameters[activeParameter].label` to distinguish the current parameter from within the signature.label. Not sure if the same applies to other language servers though.
```
{ {                                                                                                                                                                                                          
    result = {
      activeParameter = 1,
      activeSignature = 0,
      signatures = { {
          documentation = "WithTimeout returns WithDeadline(parent, time.Now().Add(timeout)).\n\nCanceling this context releases resources associated with it, so code should\ncall cancel as soon as the operations
running in this Context complete:\n\n\tfunc slowOperationWithTimeout(ctx context.Context) (Result, error) {\n\t\tctx, cancel := context.WithTimeout(ctx, 100*time.Millisecond)\n\t\tdefer cancel()  // releases resou
rces if slowOperation completes before timeout elapses\n\t\treturn slowOperation(ctx)\n\t}\n",
          label = "WithTimeout(parent context.Context, timeout time.Duration) (context.Context, context.CancelFunc)",
          parameters = { {
              label = "parent context.Context"
            }, {
              label = "timeout time.Duration"
            } }
        } }
    }
  } }
```

**Comment by urandom** [comment]
I attempted to do a prototype of how it might look and ended up with the following diff:

```diff
--- util.lua	2021-04-26 16:01:35.717186407 +0200
+++ /tmp/util.lua	2021-04-26 16:01:28.667154164 +0200
@@ -799,6 +799,25 @@
   return contents
 end
 
+local function parameter_range_in_signature(signature, active_parameter)
+  local from, to = 0, 0
+  for i = 0,#signature.parameters-1,1 do
+    local parameter_label = signature.parameters[i+1].label
+    from, to = signature.label:find(parameter_label, to + 1, true)
+
+    if from == nil then
+      return nil
+    end
+
+    if i == active_parameter then
+      return from, to
+    end
+
+  end
+
+  return nil
+end
+
 --- Converts `textDocument/SignatureHelp` response to markdown lines.
 ---
 --@param signature_help Response of `textDocument/SignatureHelp`
@@ -822,40 +841,57 @@
   if not signature then
     return
   end
-  vim.list_extend(contents, vim.split(signature.label, '\n', true))
-  if signature.documentation then
-    M.convert_input_to_markdown_lines(signature.documentation, contents)
-  end
+
+  local label = signature.label
+  local parameter
   if signature.parameters and #signature.parameters > 0 then
     local active_parameter = signature_help.activeParameter or 0
     -- If the activeParameter is not inside the valid range, then clip it.
     if active_parameter >= #signature.parameters then
       active_parameter = 0
     end
-    local parameter = signature.parameters[active_parameter + 1]
+    parameter = signature.parameters[active_parameter + 1]
     if parameter then
-      --[=[
-      --Represents a parameter of a callable-signature. A parameter can
-      --have a label and a doc-comment.
-      interface ParameterInformation {
-        --The label of this parameter information.
-        --
-        --Either a string or an inclusive start and exclusive end offsets within its containing
-        --signature label. (see SignatureInformation.label). The offsets are based on a UTF-16
-        --string representation as `Position` and `Range` does.
-        --
-        --*Note*: a label of type string should be a substring of its containing signature label.
-        --Its intended use case is to highlight the parameter label part in the `SignatureInformation.label`.
-        label: string | [number, number];
-        --The human-readable doc-comment of this parameter. Will be shown
-        --in the UI but can be omitted.
-        documentation?: string | MarkupContent;
-      }
-      --]=]
-      -- TODO highlight parameter
-      if parameter.documentation then
-        M.convert_input_to_markdown_lines(parameter.documentation, contents)
+      local from, to
+      if type(parameter.label) == "string" then
+        from, to = parameter_range_in_signature(signature, active_parameter)
+      elseif #parameter.label == 2 then
+        from = vim.str_byteindex(signature.label, parameter.label[1]+1, 1)
+        to = vim.str_byteindex(signature.label, parameter.label[2]+1, 1)
       end
+
+      if from ~= nil and to ~= nil then
+        label = label:sub(0, from-1) .. "**" .. label:sub(from, to) .. "**" .. label:sub(to+1)
+      end
+    end
+  end
+
+  vim.list_extend(contents, vim.split(label, '\n', true))
+  if signature.documentation then
+    M.convert_input_to_markdown_lines(signature.documentation, contents)
+  end
+  if parameter then
+    --[=[
+    --Represents a parameter of a callable-signature. A parameter can
+    --have a label and a doc-comment.
+    interface ParameterInformation {
+      --The label of this parameter information.
+      --
+      --Either a string or an inclusive start and exclusive end offsets within its containing
+      --signature label. (see SignatureInformation.label). The offsets are based on a UTF-16
+      --string representation as `Position` and `Range` does.
+      --
+      --*Note*: a label of type string should be a substring of its containing signature label.
+      --Its intended use case is to highlight the parameter label part in the `SignatureInformation.label`.
+      label: string | [number, number];
+      --The human-readable doc-comment of this parameter. Will be shown
+      --in the UI but can bsignature_help.signatures[active_signature+1]e omitted.
+      documentation?: string | MarkupContent;#signature.parameters-
+    }
+    --]=]
+    -- TODO highlight parameter
+    if parameter.documentation then
+      M.convert_input_to_markdown_lines(parameter.documentation, contents)
     end
   end
   return contents

```


I'm not sure whether the handling of index-based parameter labels is correct. Also, using neovim's `vim.lsp.buf.signature_help` shows the literal markdown line, including the `**` symbols. Using `ray-x/lsp_signature.nvim` on the other hand, the signature line's markdown seems to be parsed correctly, and the active parameter is shown in bold

**Comment by ray-x** [comment]
I would hope the changes will be merge soon.

**Comment by mjlbach** [comment]
Is this not a duplicate of https://github.com/neovim/neovim/issues/14217?

**Comment by urandom** [comment]
@mjlbach 

actually, from what i understand #14217 is about supporting switching between active signatures, not highlighting parameters within the currently active signature.

**Comment by mjlbach** [comment]
All that issue is about is adding support for handling the active signature result, I believe OP wanted to match the vscode handling on Reddit which includes highlighting and switching between signature items.

**Comment by mjlbach** [comment]
Folke and I talked offline about a simpler implementation, so this was closed in #15018.
